### PR TITLE
upgradable message-id

### LINKED ISSUE #6903: Replies are shown as ordinary messages if they are sent from current nightly and quoted on release build or when reply to old message sent before upgrade and both contacts are on release build
# Case 1: Replies are shown as ordinary messages if they are sent from current nightly and quoted on release build
### Description

*Type*: Bug
*Summary*: when user has current nightly installed and replies to message, sent from release build, user with release build will see an ordinary message and vice versa.

#### Expected behavior
can see reply message on both A and B

#### Actual behavior
ordinary message on device A and reply message from device B
deblkdjsf

### Reproduction

**Prerequisites:** 
1. device A has installed  0.9.31 version (released) 
2. device B has installed  latest nightly 26/11/2018

- Open Status on device A
- Log in
- Join any channel (i.e. #test)
- Open Status on device A
- Log in
- Join same channel (i.e. #test)
- Send a message from device A
- Reply to message from device B
- Check public chat on device A and B

### Additional Information
* Status version: nightly 26/11/2018
* Operating System: all

# Case 2:  No quotation is shown in the reply message on recipient side if reply to message sent before upgrade

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Bug

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
*Summary*: Recipient gets no quotation in the reply message if reply to message sent before upgrade from 0.9.31 to 0.9.32. On sender side quotation is shown as expected. Both recipient and sender are on 0.9.32. Note that if reply to message sent after upgrade is done on both sides then quotation is shown as expected for both sender and recipient. So, issue is only on recipient side for replies to "old" messages sent while being on 0.9.31.

Recipient (no quotation is shown):

Sender (shows quotation as expected)
screenshot_20181129-160441

#### Expected behavior
[comment]: # (Describe what you expected to happen.)
Quotation is shown to recipient, sender and other chat participants for replies to messages sent on 0.9.31 and on 0.9.32 
#### Actual behavior
[comment]: # (Describe what actually happened.)
Quotation is not shown to recipient if reply to message sent before upgrade
### Reproduction
[comment]: # (Describe how we can replicate the bug step by step.)

1. sender and recipient are on 0.9.31 and they are in contacts of each other
2. both send text message in 1:1 chat
3. both upgrade to 0.9.32
4. sender replies to message from recipient in 1:1 chat
5. check how reply is shown on recipient side

### Additional Information
[comment]: # (Please do your best to fill this out.)
* Status version: from 0.9.31 release to build 7 of 0.9.32 release
* Operating System: Android and iOS

### LINKED ISSUE #6956: Upgradable message-ids
# Problem

Whenever we make changes to the way message-ids are generated it creates upgradability issues. It happens again in latest release with message responses https://github.com/status-im/status-react/issues/6903#issuecomment-443127013

# Implementation

We need to pass the message-ids with the message to avoid that in the future. These are the problems I think should be taken into consideration:
- pk of the sender and chat-id might be prefixed to the message-id locally to avoid forged message-id overwriting previous messages
- chat-id needs to be consistent so the previous point might just move the problem there

# Acceptance Criteria

- message-id is speced and easily reproducible in other implementations of the communication protocol
- message-id is passed in messages with reasonable guarantees that it cannot be used to overwrite previous messages (ignoring existing ids can be considered but this will open up to MIT attacks if we don't prefix pk, an attacker could silent someone in a public chat by overwritting all their messages for a broad part of the network if they have better connectivity)

### PR DESCRIPTION
**Pull request by rasom**
fix #6956
fix #6903

- `transport.utils/message-id` function is called only in three places now and accepts `from` and  `raw_payload` as parameters. ID is calculated as `sha3(from + raw_payload)`.
- This means that for wrapped private group chat message the raw payload of `GroupMembershipUpdate` is used.
- tested on pub chats, 1-1 chats, private chats on sent/received messages
- db migrations, since `0.9.31` are merged where possible so that they will take the minimum necessary time to be executed and also compatible with nightlies builds (to not break desktop version)

status: ready

### CODE REVIEW

**Comment by rasom** [comment]
@cammellos @mandrigin please review last commit

File: src/status_im/transport/utils.cljs
Comment (mandrigin) [inline src/status_im/transport/utils.cljs]: you mean, `raw-payload`?

File: src/status_im/transport/utils.cljs
Comment (mandrigin) [inline src/status_im/transport/utils.cljs]: In which cases will we *not* have a raw payload?

Comment (mandrigin) [reply src/status_im/transport/utils.cljs]: It is worth documenting anyway

Comment (rasom) [reply src/status_im/transport/utils.cljs]: https://github.com/status-im/status-react/blob/f4690b97ae0039ac2e80d89ab4d0e8c86c700433/src/status_im/chat/models/message.cljs#L307

File: src/status_im/transport/utils.cljs
Comment (rasom) [reply src/status_im/transport/utils.cljs]: correct, will rename

Comment (rasom) [reply src/status_im/transport/utils.cljs]: fixed

File: src/status_im/transport/utils.cljs
Comment (rasom) [reply src/status_im/transport/utils.cljs]: now the only parameter is `raw-payload`, and there is a separate function for system messages ids

File: src/status_im/transport/utils.cljs
Comment (mandrigin) [reply src/status_im/transport/utils.cljs]: thanks!

File: src/status_im/transport/utils.cljs
Comment (mandrigin) [reply src/status_im/transport/utils.cljs]: awesome!

**Comment by rasom** [comment]
@cammellos @mandrigin 
- `transport.utils/message-id` function is called only in three places now and accepts raw-payload as a parameter
- tested on pub chats, 1-1 chats, private chats on sent/received messages
- db migrations, since `0.9.31` are merged where possible so that they will take the minimum necessary time to be executed and also compatible with nightlies builds (to not break desktop version)

**Comment by mandrigin** [comment]
@rasom so, seen messages and replies work between this release and the current published mobile release?

**Comment by rasom** [comment]
As far as i understand `seen` didn't work in `0.9.31`, so i didn't touch that part. But replies work between 0.9.31 and this version.

**Comment by rasom** [comment]
Though replies were in scope of https://github.com/status-im/status-react/pull/6976

**Comment by asemiankevich** [comment]
@rasom i just checked last available build for android, no quoting when upgrading from release to PR in the following case:

1. install release from playstore
2. go to public chat
3. post long message (so Read more link is shown). Note: message contains [obj] objects, most likely picture. Copy pasted this article (https://www.interfax.ru/world/640707)
4. quote this long message
5. upgrade to this PR

As result - no quotation is shown. Checking if this is in develop too so maybe not related to this PR
cc @Serhy @churik . UPD: same case in release - nightly pair works okay.

**Comment by churik** [comment]
@rasom 
On release builds replies to messages with links sent before upgrade are not shown.
*Requirements:* Device1 with release build, Device2 with release build
Steps:
- On Device1 log in and join some test chat (i.e. #test)
- On Device2 log in and join some test chat (i.e. #test)
- On Device1 send message `google.com Before upgrade`;
- On Device1 upgrade to current PR build;
- On Device1 reply to message `google.com Before upgrade`;
- On Device2 log in and check #test channel

Expected results:
can see quoting message

Actual result:
12345i

**Comment by rasom** [comment]
@asemiankevich can this be reproduced in https://github.com/status-im/status-react/pull/6976 ?

**Comment by rasom** [comment]
@churik can this be reproduced in #6976 ?

**Comment by asemiankevich** [comment]
> @asemiankevich can this be reproduced in #6976 ?

i was not abe to reproduce it in 6976

**Comment by churik** [comment]
> @churik can this be reproduced in #6976 ?

Yes, it can.

**Comment by asemiankevich** [comment]
@rasom long text messages and messages with links  that are quoted on release build disappear after upgrading to current PR

**Comment by Serhy** [comment]
In a case I quote my own message with link - after upgrade - the quoted message in reply stands as separate message. Plus, it seems in a case of quoting pure link and replying with pure-link (i.e. text message is the link without text) - quoted part splits on two messages + duplicated quote appears 

1) Install 0.9.31 on Device and create new user
2) Join any public chat (e.g. `second-crypto12`)
3) Send a message with links `status.im`
4) Reply to above message with link
5) Upgrade to latest build off this PR (https://ci.status.im/job/status-react/job/combined/job/mobile-android/9159/artifact/StatusIm-181204-142103-d744be-pr.apk / https://i.diawi.com/2pLh1Q)
6) Log in and verify quoting message

**Before upgrade:**
screenshot 2018-12-04 at 17 10 06
**After upgrade:**
screenshot 2018-12-04 at 17 10 19


<blockquote></blockquote>
<blockquote><div><strong><a href="https://i.diawi.com/2pLh1Q">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
@asemiankevich @churik @Serhy please check the last build, issues related to replies to links and long messages should be fixed https://github.com/status-im/status-react/pull/6987#issuecomment-444135660

**Comment by churik** [comment]
@rasom 
after upgrade on IOS all messages, sent before the upgrade are doubled;
sent

Steps:
- install release build
- send several messages to public chat
- upgrade to https://github.com/status-im/status-react/pull/6987#issuecomment-444135660

**Comment by rasom** [comment]
@churik tested these steps, can't reproduce

**Comment by rasom** [comment]
@churik most likely you upgraded nightly build

**Comment by churik** [comment]
@rasom rechecking

**Comment by churik** [comment]
@rasom Somehow it was problem with links in diawi (https://webapp.diawi.com/install/gdfUvR should be replaced with  https://webapp.diawi.com/install/UgwFtS for upgrade).
Not reproducible for me now, agree. Thanks for quick reply!

<blockquote><div><strong><a href="https://webapp.diawi.com/install/gdfUvR">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>
<blockquote><div><strong><a href="https://webapp.diawi.com/install/UgwFtS">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by Serhy** [comment]
It looks good for replies from my side as well. @rasom could you please resolve conflicts? 
We'll run autotest meanwhile agains latest builds.

**Comment by rasom** [comment]
@Serhy rebased

**Comment by rasom** [comment]
@mandrigin @cammellos is it fine that signature is not included in the calculation of the message id? The message which is sent by someone who intercepted the payload might be received before the original one, and for the recipient, the second person will be a valid author of the message. Should we concatenate payload with a `from` field before `sha3`?

**Comment by cammellos** [comment]
@rasom yes, we should either append the pk of the signature or the signature itself

**Comment by rasom** [comment]
So which version is preferred, `from + sha3(raw_payload)` or `sha3(from + raw_payload)`? I suppose the former has lower chances to produce the collision.

**Comment by cammellos** [comment]
`from + sha3(raw_payload)` has some drawbacks, as now a `message-id` leaks metadata if passed in a PN notification @PombeirP (the user sending the message)

`sha3(from + raw_payload)` does not have this problem, collisions I would not worry about, also this PR will likely break pairing and keeping your devices synchronized, so please do not merge before ( I don't mind testing it myself, but won't have time today)

**Comment by rasom** [comment]
@cammellos i suppose it anyway will be merged only tomorrow, np
adding `sha3(from + raw_payload)` then

**Comment by rasom** [comment]
@cammellos 
> also this PR will likely break pairing and keeping your devices synchronized, so please do not merge before ( I don't mind testing it myself, but won't have time today)

is it planned to be included to 0.9.32? I mean pairing. Also, should it be tested in this PR right now or...?

**Comment by cammellos** [comment]
@rasom it will be, but I think it's actually fine, I will quickly test it, latest build has all the required changes already?

**Comment by cammellos** [comment]
@rasom tested https://github.com/status-im/status-react/pull/6987#issuecomment-444183808 pairing/group chats and it's all good on my side

**Comment by rasom** [comment]
both commits are cherry-picked onto `0.9.32`

**Comment by asemiankevich** [comment]
@cammellos @yenda please add your reviews here

File: src/status_im/data_store/realm/schemas/account/migrations.cljs
Comment (cammellos) [inline src/status_im/data_store/realm/schemas/account/migrations.cljs:162]: This is ok to remove in develop as well?

File: src/status_im/data_store/realm/schemas/account/migrations.cljs
Comment (cammellos) [inline src/status_im/data_store/realm/schemas/account/migrations.cljs:179]: I take this is a copy & paste of the code from somewhere else, if that's the case, could you add a comment begin/end of copy and paste code?

**Review by cammellos — APPROVED** [review]

File: src/status_im/data_store/realm/schemas/account/migrations.cljs
Comment (rasom) [reply src/status_im/data_store/realm/schemas/account/migrations.cljs:162]: yes, it is ok

File: src/status_im/data_store/realm/schemas/account/migrations.cljs
Comment (rasom) [reply src/status_im/data_store/realm/schemas/account/migrations.cljs:179]: some comments are added, let me know if it's not sufficient enough

**Comment by churik** [comment]
@rasom I can see commits from this PR in [release job](https://ci.status.im/job/status-react/job/release/job/release%252F0.9.32/22/) - should we close this PR or waiting for another review to merge it to develop?

**Comment by rasom** [comment]
we are waiting for another review

**Review by mandrigin — APPROVED** [review]

**Pull request merged**
